File: C++/structs/stack/stack.cpp
#include "stack.h"
#include <sstream>

using namespace std;

Stack::Stack() : top(nullptr) {}

Stack::~Stack() {
    while (top != nullptr) {
        Node* temp = top;
        top = top->next;
        delete temp;
    }
}

void Stack::push(const string& value) {
    Node* newNode = new Node(value);
    newNode->next = top;
    top = newNode;
}

bool Stack::pop(string& value) {
    if (top == nullptr) {
        return false;
    }
    
    value = top->value;
    Node* temp = top;
    top = top->next;
    delete temp;
    return true;
}

bool Stack::peek(string& value) const {
    if (top == nullptr) {
        return false;
    }
    value = top->value;
    return true;
}

bool Stack::isEmpty() const {
    return top == nullptr;
}

void Stack::printAll() const {
    if (top == nullptr) {
        cout << "[]" << endl;
        return;
    }
    
    cout << "[";
    Node* current = top;
    bool first = true;
    while (current != nullptr) {
        if (!first) cout << ", ";
        cout << current->value;
        first = false;
        current = current->next;
    }
    cout << "]" << endl;
}

string Stack::toString() const {
    if (top == nullptr) {
        return "[]";
    }
    
    ostringstream oss;
    oss << "[";
    Node* current = top;
    bool first = true;
    while (current != nullptr) {
        if (!first) oss << ", ";
        oss << current->value;
        first = false;
        current = current->next;
    }
    oss << "]";
    return oss.str();
}

File: go/structs/stack/stack.go
package stack

import "fmt"

type Node struct {
	value string
	next  *Node
}

type Stack struct {
	top *Node
}

func (s *Stack) Push(value string) {
	newNode := &Node{
		value: value,
		next:  s.top,
	}
	s.top = newNode
}

func (s *Stack) Pop() (string, bool) {
	if s.top == nil {
		return "", false
	}

	value := s.top.value
	s.top = s.top.next
	return value, true
}

func (s *Stack) Peek() (string, bool) {
	if s.top == nil {
		return "", false
	}
	return s.top.value, true
}

func (s *Stack) IsEmpty() bool {
	return s.top == nil
}

func (s *Stack) PrintAll() {
	if s.top == nil {
		fmt.Println("[]")
		return
	}

	var values []string
	current := s.top
	for current != nil {
		values = append([]string{current.value}, values...)
		current = current.next
	}
	fmt.Println(values)
}

func (s *Stack) String() string {
	if s.top == nil {
		return "[]"
	}

	var values []string
	current := s.top
	for current != nil {
		values = append([]string{current.value}, values...)
		current = current.next
	}
	return fmt.Sprint(values)
}
